e2e_playwright/multipage_apps_v2/mpa_v2_basics_test.py

```diff
@@ -0,0 +1,336 @@
+    """Test that we can navigate to a page directly via URL. using the title."""
+    page.goto(f"http://localhost:{app_port}/Different_Title")
+    wait_for_app_loaded(page)
+
+    expect(page_heading(page)).to_contain_text("Page 3")
+
+
+def test_can_switch_between_pages_and_edit_widgets(app: Page):
+    """Test that we can switch between pages and page widgets do not persist."""
+    get_page_link(app, "Different Title").click()
+    wait_for_app_run(app, wait_delay=1000)
+
+    slider = app.locator('.stSlider [role="slider"]').nth(1)
+    slider.click()
+    slider.press("ArrowRight")
+    wait_for_app_run(app)
+    expect(app.get_by_test_id("stMarkdown").nth(1)).to_contain_text("x is 1")
+
+    get_page_link(app, "page 2").click()
+    wait_for_app_run(app, wait_delay=1000)
+
+    get_page_link(app, "Different Title").click()
+    wait_for_app_run(app, wait_delay=1000)
+
+    expect(app.get_by_test_id("stMarkdown").nth(1)).to_contain_text("x is 0")
+
+
+def test_titles_are_set_correctly(app: Page):
+    """Test that page titles work as expected"""
+    expect_page_order(app)
+
+
+def test_dynamic_pages(themed_app: Page, assert_snapshot: ImageCompareFunction):
+    """Test that dynamic pages are defined"""
+    check_field(themed_app, dynamic_pages=True)
+    wait_for_app_run(themed_app)
+
+    nav = themed_app.get_by_test_id("stSidebarNav")
+    expect_page_order(themed_app, ["page 2", "Different Title", "page 5", "page 9"])
+
+    assert_snapshot(nav, name="dynamic-pages")
+
+
+def test_show_not_found_dialog(page: Page, app_port: int):
+    """Test that we show a not found dialog if the page doesn't exist."""
+    page.goto(f"http://localhost:{app_port}/not_a_page")
+    wait_for_app_loaded(page)
+
+    expect(page.locator('[role="dialog"]')).to_contain_text("Page not found")
+
+
+def test_handles_expand_collapse_of_mpa_nav_correctly(
+    themed_app: Page, assert_snapshot: ImageCompareFunction
+):
+    """Test that we handle expand/collapse of MPA nav correctly."""
+
+    check_field(themed_app, add_sidebar_elements=True)
+    wait_for_app_run(themed_app)
+
+    view_button = themed_app.get_by_test_id("stSidebarNavViewButton")
+
+    expect(view_button).to_be_visible()
+
+    # Expand the nav
+    view_button.click(force=True)
+    # We apply a quick timeout here so that the UI has some time to
+    # adjust for the screenshot after the click
+    themed_app.wait_for_timeout(250)
+    assert_snapshot(
+        themed_app.get_by_test_id("stSidebarNav"), name="mpa-sidebar_nav_expanded"
+    )
+
+    # Collapse the nav
+    view_button.click(force=True)
+    themed_app.wait_for_timeout(250)
+    assert_snapshot(
+        themed_app.get_by_test_id("stSidebarNav"), name="mpa-sidebar_nav_collapsed"
+    )
+
+    # Expand the nav again
+    view_button.click(force=True)
+    themed_app.wait_for_timeout(250)
+    assert_snapshot(
+        themed_app.get_by_test_id("stSidebarNav"), name="mpa-sidebar_nav_expanded"
+    )
```

Comment:
nit: could we wait on some other thing such as expected pages appearing at the bottom of the list being present/absent so that we don't have to have a hard coded number here?